trie/zk_trie_impl.go

```diff
@@ -72,14 +82,23 @@ func NewZkTrieImplWithRoot(storage ZktrieDatabase, root *zkt.Hash, maxLevels int
 }
 
 // Root returns the MerkleRoot
-func (mt *ZkTrieImpl) Root() *zkt.Hash {
+func (mt *ZkTrieImpl) Root() (*zkt.Hash, error) {
+	hashedDirtyStorage := make(map[zkt.Hash]*Node)
+	rootKey, err := mt.commit(mt.rootKey, hashedDirtyStorage, new(sync.Mutex))
+	if err != nil {
+		return nil, err
+	}
+
+	mt.rootKey = rootKey
+	mt.dirtyIndex = big.NewInt(0)
+	mt.dirtyStorage = hashedDirtyStorage
```

Comment:
I don't think the `hashedDirtyStorage` should replace the `dirtyStorage` here since `hashedDirtyStorage` is an index of nodes which have not been "dirty" (i.e. their node hash has been calculated). With this update, we notice that if we have called `commit` right after a `root` call, we have still need to recalculate all node hashes again (they should have been calculated in the `root` calling before).

To cache the calculated hashes we should maintaining a "cahce" index from the "dirty" node key to the "real" (calculated) node hash. In `Root()`, for each calculating of nodehash, we search the cache index first. If the corresponding entry is not found, we calculate the node hash and save it into the cache index. And for `dirtyStorage` table, all the key inside it should be generated by `newDirtyNodeKey` (and I advice assign 0 for root node(

Reply:
> we have still need to recalculate all node hashes

that is not really the case since nodes in `hashedDirtyNodes` already have their hashes calculated and cached in their nodeHash field.